arrow/arrow.py

```diff
@@ -1390,84 +1497,101 @@ def for_json(self):
+    def __radd__(self, other: Union[timedelta, relativedelta]) -> "Arrow":
         return self.__add__(other)
 
-    def __sub__(self, other):
+    @overload
+    def __sub__(self, other: Union[timedelta, relativedelta]) -> "Arrow":
+        pass
+
+    @overload
+    def __sub__(self, other: Union[dt_datetime, "Arrow"]) -> timedelta:
+        pass
+
+    def __sub__(
+        self, other: Union[timedelta, relativedelta, dt_datetime, "Arrow"]
+    ) -> Union["Arrow", timedelta]:
 
         if isinstance(other, (timedelta, relativedelta)):
             return self.fromdatetime(self._datetime - other, self._datetime.tzinfo)
 
-        elif isinstance(other, datetime):
+        elif isinstance(other, dt_datetime):
             return self._datetime - other
 
         elif isinstance(other, Arrow):
             return self._datetime - other._datetime
 
-        return NotImplemented
+        # FIXME: change to raise
+        return NotImplemented  # type: ignore
 
-    def __rsub__(self, other):
+    def __rsub__(self, other: dt_datetime) -> timedelta:
 
-        if isinstance(other, datetime):
+        if isinstance(other, dt_datetime):
             return other - self._datetime
 
-        return NotImplemented
+        # FIXME: change to raise
+        return NotImplemented  # type: ignore
 
     # comparisons
 
-    def __eq__(self, other):
+    def __eq__(self, other: Any) -> bool:
 
-        if not isinstance(other, (Arrow, datetime)):
+        if not isinstance(other, (Arrow, dt_datetime)):
             return False
 
         return self._datetime == self._get_datetime(other)
 
-    def __ne__(self, other):
+    def __ne__(self, other: Any) -> bool:
 
-        if not isinstance(other, (Arrow, datetime)):
+        if not isinstance(other, (Arrow, dt_datetime)):
             return True
 
         return not self.__eq__(other)
 
-    def __gt__(self, other):
+    def __gt__(self, other: Union["Arrow", dt_datetime]) -> bool:
 
-        if not isinstance(other, (Arrow, datetime)):
-            return NotImplemented
+        if not isinstance(other, (Arrow, dt_datetime)):
+            # FIXME: change to raise
+            return NotImplemented  # type: ignore
 
         return self._datetime > self._get_datetime(other)
 
-    def __ge__(self, other):
+    def __ge__(self, other: Union["Arrow", dt_datetime]) -> bool:
 
-        if not isinstance(other, (Arrow, datetime)):
-            return NotImplemented
+        if not isinstance(other, (Arrow, dt_datetime)):
+            # FIXME: change to raise
+            return NotImplemented  # type: ignore
 
         return self._datetime >= self._get_datetime(other)
 
-    def __lt__(self, other):
+    def __lt__(self, other: Union["Arrow", dt_datetime]) -> bool:
 
-        if not isinstance(other, (Arrow, datetime)):
-            return NotImplemented
+        if not isinstance(other, (Arrow, dt_datetime)):
+            # FIXME: change to raise
+            return NotImplemented  # type: ignore
```

Comment:
why this method returns `NotImplemented` class? I think we should raise `NotImplemented`. or I should change return type to `Union["Arrow", Type[NotImplemented]]`.